### DISCUSSION TITLE
Add the proxy connection type.

### DISCUSSION THREAD

**Discussion by langyo** [SeaQL/sea-orm#1875] [Ideas]
I think this framework should include a new connection type designed for proxy connections.

## Motivation

It is primarily applicable in the following scenarios:

- Providing support for microservices, such as WASM containers, that are containerized and do not have direct network connectivity permissions.#1348
- Offering the possibility of customization for the various peculiar interfaces provided by different cloud service providers. #466

## Proposed Solutions

To achieve these goals, I initially made the following attempts:

- I attempted to replace the dependency library ```ring``` used for initiating secure connections with ```rustls``` with ```ring-wasi``` to support the ```wasi``` compilation target. However, after trying various methods, ```rustls``` still couldn't compile successfully. This path is not viable in the short term because cryptographic verification takes a significant amount of time. Rewriting a pure Rust version (or a version that doesn't require cross-compilation) of ```ring``` poses significant security risks.

- I attempted to rewrite the existing ```mock``` database connection objects in the current library by implementing a "bridge" that can be used to forward database read and write operations through re-implementing the ```MockDatabaseTrait```. However, during my actual implementation, I realized that re-implementing it would require making some originally private objects in the library public, and it would be more efficient to create a new one instead of modifying object visibility.

My current preliminary idea is to add a new feature similar to ```mock``` called ```proxy```, which can simultaneously provide support for all three databases originally supported by ```sqlx```: ```proxy-mysql```, ```proxy-postgresql```, and ```proxy-sqlite```. I plan to create three new connection types similar to the original three connection types corresponding to ```mock```.

## Current Workarounds

At the moment, I haven't created a PR for this. I'm considering implementing this functionality in the coming days to meet my recent development needs. If everyone finds it feasible, I will attempt to write it in the next couple of days.

**Comment by tyt2y3** [comment]
What are those proxies `proxy-mysql` etc look like? Are there some existing projects?

**Comment by langyo** [comment]
> What are those proxies `proxy-mysql` etc look like? Are there some existing projects?

There isn't a project for this at the moment since the interface hasn't been created yet. However, I can provide some pseudocode to outline my ideas:

```rust
use sea_orm::*;

fn query(stmt: Statement) -> Result<Vec<QueryResult>, DbErr> {
  // Own implementation
}

fn execute(stmt: Statement) -> Result<ExecResult, DbErr> {
  // Own implementation
}

let db = ProxyDatabase::connect(ProxyDatabaseConfig {
  query, execute
}).await?;
let builder = db.get_database_backend();

// Create table
db.execute(
    builder.build(
        Schema::new(builder)
            .create_table_from_entity(models::test::Entity)
            .if_not_exists(),
    ),
)
.await?;

// Query test
let test = models::test::Entity::find().all(&db).await?.to_vec();

// Insert test
models::test::Entity::insert(models::test::ActiveModel {
    name: Set("test".to_owned()),
    ..Default::default()
}).exec(&db).await?;
```

**Comment by tyt2y3** [comment]
This is an interesting idea but this will require a pretty big architectural change.

**Reply by langyo** [reply]
Essentially, this function is to implement another version of ```mock``` database connection, but it changes the fixed test data to the user forwarding the request and results, and communicates between the two parties by itself.

I think I can try to change from ```MockDatabase```.

**Comment by langyo** [comment]
Reference #1881
